pinot-core/src/main/java/org/apache/pinot/core/operator/combine/GroupByOrderByCombineOperator.java

```diff
@@ -125,43 +123,44 @@ public String getOperatorName() {
    */
   @Override
   protected void processSegments(int threadIndex) {
-    try {
-      IntermediateResultsBlock intermediateResultsBlock =
-          (IntermediateResultsBlock) _operators.get(threadIndex).nextBlock();
-
-      _initLock.lock();
-      try {
-        if (_dataSchema == null) {
-          _dataSchema = intermediateResultsBlock.getDataSchema();
-          // NOTE: Use trimSize as resultSize on server size.
-          if (_trimThreshold >= MAX_TRIM_THRESHOLD) {
-            // special case of trim threshold where it is set to max value.
-            // there won't be any trimming during upsert in this case.
-            // thus we can avoid the overhead of read-lock and write-lock
-            // in the upsert method.
-            _indexedTable = new UnboundedConcurrentIndexedTable(_dataSchema, _queryContext, _trimSize);
-          } else {
-            _indexedTable =
-                new ConcurrentIndexedTable(_dataSchema, _queryContext, _trimSize, _trimSize, _trimThreshold);
+    for (int operatorIndex = threadIndex; operatorIndex < _numOperators; operatorIndex += _numThreads) {
+      IntermediateResultsBlock resultsBlock = (IntermediateResultsBlock) _operators.get(operatorIndex).nextBlock();
+
+      if (_indexedTable == null) {
+        _initLock.lock();
+        try {
+          if (_indexedTable == null) {
```

Comment:
Because `_indexedTable` isn't volatile, double checked locking doesn't work. The JMM allows reordering of the checks and lock.

Reply:
Lock will enforce the memory synchronization, thus guarantee that the inner check getting the latest value. I don't think JVM can reorder the check and lock because of the memory barrier.
More can be read here under the `Memory Synchronization` section: https://docs.oracle.com/javase/1.5.0/docs/api/java/util/concurrent/locks/Lock.html

Reply:
Lock acquisition is guaranteed to happen before the second check, but the first check can be reordered with the lock acquisition, which nullifies the benefits of the idiom. 

Since `_indexedTable` isn't immutable, there is also the issue of unsafe publication. It's worth taking a look here, where the example uses `synchronized` but applies equally to `ReentrantLock`: https://shipilev.net/blog/2014/safe-public-construction/

Making `_indexedTable` makes the problem go away though.

Reply:
Having said that, `IndexedTable` is _mostly_ immutable so it might be ok in practice https://github.com/apache/pinot/blob/f3068bc93165e359cc3d846c40f62312ecc98af3/pinot-core/src/main/java/org/apache/pinot/core/data/table/IndexedTable.java#L39L46

Reply:
Thanks for sharing this article. Learned a lot from it! Changed it to `volatile`